rfcs/select-component.md

```diff
@@ -0,0 +1,191 @@
+# Select Component
+
+### Exports
+
+* `StatefulSelect`
+* `StatefulSelectContainer`
+* `Select`
+* `StyledSelectInput`
+* `StyledOption`
+* `StyledDropDown`
+* `StyledLabel`
+* `StyledHint`
+
+### `Select` and `StatefulSelect` API
+
+* `type: TYPE.SEARCH | TYPE.SELECT`:
+  type of component to be in select or search mode (with lookup input)
+* `selectedOptions: Array<Object>`:
+  Current selected options. Every option object has `id: string` and `label: React$Node`. Label is defaulted to display for selected option, otherwise see `getSelectedOptionLabel` method
+* `placeholder: ?string`:
+  Placeholder text if nothing is selected. Default is `Choose one...`
+* `rows: ?number`:
+  Represents maximum visible length of options, all other will be scrolled. If not defined, all options will be visible.
+* `options: Array<Object>`:
+  All Options in dropdown. Should be provided for Select and Search mode equally. Every option object has `id: string` and `label: React$Node`. Label is defaulted to display for option in dropdown, otherwise see `getOptionLabel` method. Optional `disabled: boolean` for option to be disabled from selection.
+* `label: ?React$Node`:
+  Component or String representing label for Select component. Default is `''`
+* `caption: ?React$Node`:
+  Component or String representing hint message for Select component. Default is `''`
+* `textValue: ?string`:
+  Initial text value for Search to lookup for text.
+* `error: ?string`:
+  Error message and error state. Default is `''`
+* `multiple: ?boolean`:
+  Sets if multiple choices are allowed in Select component. Default is `false`
+* `autoFocus: boolean`:
+  make the control focused (active). Default is `false`
+* `disabled: boolean`:
+  Disable control from being changed
+* `required: boolean`:
+  Mark control as required
+* `overrides: {}`
+  * `DropDown: ?React$Node` component to use for dropdown list
+  * `Option: ?React$Node` component to use for options in dropdown list
+  * `Root: ?React$Node` component to use for most top of the select component
+  * `Input: ?React$Node` component for Input showing current selected value(s). See `Input` Control of this framework for reference to override it's functionality.
+  * `SearchIcon: ?React$Node` component for all icons appearing in Select component. It's provided `$type: ICON.LOOP | ICON.CLEAR_TAG | ICON.CLEAR_ALL | ICON.SELECTED` to setup corresponding icon of Select component
```

Comment:
I kind of think each of these icons should be individually override-able. If I just want to override the CLEAR_TAG, I don't want to have to wire up a bunch of defaults for the other three icons.

Reply:
there is only one prop to override for icon ($type)
Otherwise we would have to add 5 more different components just to support all of them.
With `override` approach it would be harder to change code everywhere and not mistaken if we wish to maintain or add changes here